bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 	return ctrl.Result{}, nil
 }
 
+// resolveProxyUrl adds the user information to the proxy url
+func resolveProxyUrl(proxyUrl string, secret *corev1.Secret) (string, error) {
+	if secret == nil {
+		return proxyUrl, nil
+	}
+	username := secret.StringData["username"]
+	password := secret.StringData["password"]
+	parsedUrl, err := url.Parse(proxyUrl)
+	if err != nil {
+		return "", err
+	}
+	parsedUrl.User = url.UserPassword(username, password)
+	return parsedUrl.String(), nil
+}
+
+// resolveProxyFiles resolves the required proxy unit files to cloudinit.Files
+func (r *KubeadmConfigReconciler) resolveProxyFiles(ctx context.Context, cfg *bootstrapv1.KubeadmConfig) ([]bootstrapv1.File, error) {
+	collected := make([]bootstrapv1.File, 0)
+
+	if cfg.Spec.Proxy == nil {
+		return nil, nil
+	}
+
+	// resolve credentials
+	var httpAuth, httpsAuth *corev1.Secret
+	if cfg.Spec.Proxy.HttpProxyAuth != nil {
+		httpAuth = &corev1.Secret{}
+		key := types.NamespacedName{Namespace: cfg.Namespace, Name: *cfg.Spec.Proxy.HttpProxyAuth}
+		if err := r.Client.Get(ctx, key, httpAuth); err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, errors.Wrapf(err, "secret not found: %s", key)
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve Secret %q", key)
+		}
+	}
+	if cfg.Spec.Proxy.HttpsProxyAuth != nil {
+		httpsAuth = &corev1.Secret{}
+		key := types.NamespacedName{Namespace: cfg.Namespace, Name: *cfg.Spec.Proxy.HttpsProxyAuth}
+		if err := r.Client.Get(ctx, key, httpsAuth); err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, errors.Wrapf(err, "secret not found: %s", key)
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve Secret %q", key)
+		}
+	}
+	// resolve urls
+	httpProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpProxy, httpAuth)
+	if err != nil {
+		return nil, err
+	}
+	httpsProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpProxy, httpAuth)
```

Comment:
I guess this should be
```suggestion
	httpsProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpsProxy, httpsAuth)
```